builder/vsphere/common/step_keep_cdrom.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+//go:generate packer-sdc struct-markdown
+//go:generate packer-sdc mapstructure-to-hcl2 -type KeepCDRomConfig
+
+package common
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/packer-plugin-sdk/multistep"
+	packersdk "github.com/hashicorp/packer-plugin-sdk/packer"
+	"github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/driver"
+)
+
+type KeepCDRomConfig struct {
+	// Keep a number of CD-ROM devices. Range: 1-4.
+	// If set to 0, `keep_cdrom` is ignored and the step is skipped.
+	// When set to a value in the range, `remove_cdrom` is ignored and
+	// the CD-ROM devices are kept without any attached media.
+	KeepCdrom int32 `mapstructure:"keep_cdrom"`
+}
+
+type StepKeepCDRom struct {
+	Config      *KeepCDRomConfig
+	CDRomConfig *CDRomConfig
+}
+
+func (s *StepKeepCDRom) Run(_ context.Context, state multistep.StateBag) multistep.StepAction {
+	ui := state.Get("ui").(packersdk.Ui)
+	vm := state.Get("vm").(driver.VirtualMachine)
+
+	var err error
+
+	// Check if `keep_cdrom` is set.
+	keepCdrom := s.Config.KeepCdrom
+	if keepCdrom > 0 {
+		if keepCdrom < 1 || keepCdrom > 4 {
+			// If `keep_cdrom` is set, it must be between 1 and 4.
+			err := fmt.Errorf("'keep_cdrom' should be between 1 and 4")
+
+			// Add the error to the state bag and halt the execution.
+			state.Put("error", fmt.Errorf("error keeping cdrom: %v", err))
+			return multistep.ActionHalt
+		}
+	}
+
+	// Keep the CD-ROM devices on the image without any attached media.
+	if keepCdrom > 0 {
+		ui.Say("Resetting CD-ROM device...")
+		// Remove all CD-ROM devices from the image.
+		err := vm.RemoveCdroms()
+		if err != nil {
+			state.Put("error", fmt.Errorf("error removing cdrom: %v", err))
+			return multistep.ActionHalt
+		}
+		// If the CD-ROM device type is SATA, find the SATA controller.
+		if s.CDRomConfig.CdromType == "sata" {
+			if _, err := vm.FindSATAController(); err == driver.ErrNoSataController {
+				ui.Say("Adding SATA controller...")
+				// If a SATA controller is not found, add it.
+				if err := vm.AddSATAController(); err != nil {
+					state.Put("error", fmt.Errorf("error adding sata controller: %v", err))
+					return multistep.ActionHalt
+				}
+			}
+		}
+	}
+
+	ui.Say("Adding CD-ROM device...")
+	// Add the CD-ROM devices to the image based on the value of `keep_cdrom`.
+	// A valid ISO path is required for this step. The media will subsequently be ejected.
+	if int(keepCdrom) > 0 {
+		for i := 0; i < int(keepCdrom); i++ {
+			err = vm.AddCdrom(s.CDRomConfig.CdromType, s.CDRomConfig.ISOPaths[0])
+			if err != nil {
+				state.Put("error", fmt.Errorf("error adding cdrom %d: %v", i, err))
+				return multistep.ActionHalt
+			}
+		}
+	}
```

Comment:
Same thing here. If we get here we know keepCdrom is between 1 and 4 so no need to check if it is greater than 0. 
```go

		for i := 0; i < int(keepCdrom); i++ {
			err = vm.AddCdrom(s.CDRomConfig.CdromType, s.CDRomConfig.ISOPaths[0])
			if err != nil {
				state.Put("error", fmt.Errorf("error adding cdrom %d: %v", i, err))
				return multistep.ActionHalt
			}
		}
```

Reply:
Resolved in 011dd5a.